trigger/process.go

```diff
@@ -72,3 +79,192 @@ func NotificationsFromEvent(e *event.EventLogEntry) ([]notification.Notification
+			if trigger.Level != model.ProjectTriggerLevelTask {
+				continue
+			}
+			if trigger.Project != t.Project {
+				continue
+			}
+			if trigger.Status != "" && trigger.Status != t.Status {
+				continue
+			}
+			if trigger.TaskRegex != "" {
+				regex, err := regexp.Compile(trigger.TaskRegex)
+				if err != nil {
+					catcher.Add(err)
+					continue
+				}
+				if !regex.MatchString(t.DisplayName) {
+					continue
+				}
+			}
+			if trigger.BuildVariantRegex != "" {
+				regex, err := regexp.Compile(trigger.BuildVariantRegex)
+				if err != nil {
+					catcher.Add(err)
+					continue
+				}
+				if !regex.MatchString(t.BuildVariant) {
+					continue
+				}
+			}
+
+			args := ProcessorArgs{
+				SourceVersion:     sourceVersion,
+				DownstreamProject: ref,
+				File:              trigger.ConfigFile,
+				Command:           trigger.Command,
+				TriggerType:       model.ProjectTriggerLevelTask,
+				TriggerID:         t.Id,
+				EventID:           e.ID,
+				DefinitionID:      trigger.DefinitionID,
+			}
+			v, err := processor(args)
+			if err != nil {
+				catcher.Add(err)
+				continue
+			}
+			if v != nil {
+				versions = append(versions, *v)
+			}
+			continue projectLoop
+		}
+	}
+
+	return versions, catcher.Resolve()
+}
+
+func triggerDownstreamProjectsForBuild(b *build.Build, e *event.EventLogEntry, processor projectProcessor) ([]version.Version, error) {
+	if b.Requester != evergreen.RepotrackerVersionRequester {
+		return nil, nil
+	}
+	downstreamProjects, err := model.FindDownstreamProjects(b.Project)
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding project ref")
+	}
+	sourceVersion, err := version.FindOneId(b.Version)
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding version")
+	}
+
+	catcher := grip.NewBasicCatcher()
+	versions := []version.Version{}
+projectLoop:
+	for _, ref := range downstreamProjects {
+
+		for _, trigger := range ref.Triggers {
+
+			if trigger.Level != model.ProjectTriggerLevelBuild {
+				continue
+			}
+			if trigger.Project != b.Project {
+				continue
+			}
+			if trigger.Status != "" && trigger.Status != b.Status {
+				continue
+			}
+			if trigger.BuildVariantRegex != "" {
+				regex, err := regexp.Compile(trigger.BuildVariantRegex)
+				if err != nil {
+					catcher.Add(err)
+					continue
+				}
+				if !regex.MatchString(b.BuildVariant) {
+					continue
+				}
+			}
+
+			args := ProcessorArgs{
+				SourceVersion:     sourceVersion,
+				DownstreamProject: ref,
+				File:              trigger.ConfigFile,
+				Command:           trigger.Command,
+				TriggerType:       model.ProjectTriggerLevelBuild,
+				TriggerID:         b.Id,
+				EventID:           e.ID,
+				DefinitionID:      trigger.DefinitionID,
+			}
+			v, err := processor(args)
+			if err != nil {
+				catcher.Add(err)
+				continue
+			}
+			if v != nil {
+				versions = append(versions, *v)
+			}
+			continue projectLoop
+		}
+	}
+
+	return versions, catcher.Resolve()
```

Comment:
is there a valid case for returnning an error and a version?

Reply:
I wanted continue on error behavior between versions, so that potentially some but not all versions are created (the others will have errors)